### PR TITLE
Performance of bind rows(.id =)

### PR DESCRIPTION
**Pull request by romainfrancois**
``` r
library(dplyr)
library(purrr)

make_df = function(x){
  nrows = sample(1:1000,1)
  df = data.frame(id = seq(1,nrows), val = rep("character",nrows), stringsAsFactors = FALSE)
  return(df)
}

make_list = function(n){
  ll = lapply(1:n, function(x) make_df(x))
  names(ll) = as.character(1:n)
  return(ll)
}

bench_list <- function(n){
  data <- make_list(n)
  sizes <- map_int(data, nrow)
  bench::mark(check = FALSE, 
    bind_rows(data), 
    bind_rows(data, .id = "id"), 
    rep(names(data), sizes)
  )
}

bench_list(1e3)
#> # A tibble: 3 x 10
#>   expression    min    mean  median     max `itr/sec` mem_alloc  n_gc n_itr
#>   <chr>      <bch:> <bch:t> <bch:t> <bch:t>     <dbl> <bch:byt> <dbl> <int>
#> 1 bind_rows… 7.04ms  8.03ms  7.87ms 10.23ms     125.     5.87MB    15    38
#> 2 "bind_row… 9.97ms 11.26ms  11.1ms 14.43ms      88.8    9.76MB    18    16
#> 3 rep(names… 5.06ms  5.61ms  5.59ms  7.09ms     178.     3.86MB    10    71
#> # … with 1 more variable: total_time <bch:tm>
bench_list(1e4)
#> # A tibble: 3 x 10
#>   expression     min    mean  median     max `itr/sec` mem_alloc  n_gc
#>   <chr>      <bch:t> <bch:t> <bch:t> <bch:t>     <dbl> <bch:byt> <dbl>
#> 1 bind_rows… 101.7ms 103.9ms 103.9ms 106.2ms      9.62      58MB     3
#> 2 "bind_row… 137.7ms 137.7ms 137.7ms 137.7ms      7.26    96.7MB     3
#> 3 rep(names…  55.3ms  59.4ms  56.6ms  73.7ms     16.8     38.5MB     2
#> # … with 2 more variables: n_itr <int>, total_time <bch:tm>
bench_list(1e5)
#> Warning: Some expressions had a GC in every iteration; so filtering is
#> disabled.
#> # A tibble: 3 x 10
#>   expression      min     mean   median      max `itr/sec` mem_alloc  n_gc
#>   <chr>      <bch:tm> <bch:tm> <bch:tm> <bch:tm>     <dbl> <bch:byt> <dbl>
#> 1 bind_rows…    2.75s    2.75s    2.75s    2.75s     0.364     575MB     1
#> 2 "bind_row…    2.68s    2.68s    2.68s    2.68s     0.373     958MB     4
#> 3 rep(names… 739.82ms 739.82ms 739.82ms 739.82ms     1.35      382MB     0
#> # … with 2 more variables: n_itr <int>, total_time <bch:tm>
```

<sup>Created on 2018-12-14 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1.9000)</sup>

### CODE REVIEW

**Comment by romainfrancois** [comment]
closes #3787

**Comment by romainfrancois** [comment]
This used to pay for too many write barrier checks and preserve/release of each `Rcpp::String` in the vector.

**Comment by romainfrancois** [comment]
There is still obviously a difference because it's more work to make the id column than not to, see the `rep(names(data), sizes)` line in the bench results.

File: src/bind.cpp [hunk 1]
```diff
@@ -295,14 +289,37 @@ List rbind__impl(List dots, const SymbolString& id) {
...
+
+      // we id_col now, so it is definitely younger than dots_names
+      // this is surely write barrier proof
+      SEXP id_col = PROTECT(Rf_allocVector(STRSXP, n));
+
+      SEXP* p_id_col = STRING_PTR(id_col);
+      for (int i = 0; i < ndata; ++i, ++p_dots_names, ++p_dots) {
+
+        // skip NULL on dots. because the way df_nrows is made above
+        // need to skip dots_names too
```
Comment (krlmlr) [inline src/bind.cpp]: This does look brittle. Can we add a sentinel `0` to `df_nrows` in this situation?

Comment (romainfrancois) [reply src/bind.cpp]: I don't understand what you mean

Comment (romainfrancois) [reply src/bind.cpp]: Ok, I see, that's because the code above does that; 

```

Comment (krlmlr) [reply src/bind.cpp]: Yes, if we replace the `continue` with a code that adds sentinels?

File: src/bind.cpp [hunk 2]
```diff
@@ -182,11 +182,9 @@ List rbind__impl(List dots, const SymbolString& id) {
   R_xlen_t n = 0;
   std::vector<SEXP> chunks;
   std::vector<R_xlen_t> df_nrows;
-  std::vector<String> dots_names;
```
[the remark this answers is not in the document]
Comment (romainfrancois) [reply src/bind.cpp]: We treat this:
 
```
SEXP dots_names(vec_names(dots));
```

That can be: 
 - NULL, in that case all names are ""
 - a character vector that is already protected as it comes from R and is older than the vector we are making, so there's no need to pay for the gatekeeping of `STRING_ELT()` over and over again, we can just copy the pointers

[the remark this answers is not in the document]
Comment (romainfrancois) [reply src/bind.cpp]: In any case, there's no need to have this in a `std::vector<>`, and especially a `std::vector<Rcpp::String>`

File: src/bind.cpp (hunk 1 above)
Comment (romainfrancois) [reply src/bind.cpp]: Maybe that loop is only needed to calculate `n` in that case. Why not.

**Pull request merged**